haskell/test/SMT/Common/Types.hs

```diff
@@ -0,0 +1,98 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+
+module SMT.Common.Types
+  ( PropertyTestArg (..)
+  , PropertyTestInput (..)
+  , MkPropertyTestInput
+  , MkGenPropose
+  , MkGenCustomCalls
+  , MkPropertyTestOption (..)
+
+  , GeneratorState (..)
+  , GeneratorT
+  , initGeneratorState
+  , runGeneratorT
+  , (<&&>)
+  ) where
+
+import Universum hiding (show)
+
+import Hedgehog
+
+import Lorentz hiding (not, cast)
+
+import Text.Show (show)
+import Tezos.Crypto (SecretKey)
+
+import Ligo.BaseDAO.Types
+import SMT.Model.BaseDAO.Types
+
+
+-- | The input of an SMT property test run
+newtype PropertyTestInput =
+  PropertyTestInput ([ModelCall], ModelState)
+  deriving stock Show
+
+data PropertyTestArg = PropertyTestArg
+  { ptaGuardianAddr :: Address -- Used in `genPropose`, `genStorage`
+  , ptaGovAddr :: Address -- Used in `genPropose`, `genTransferContractTokens`
+  , ptaViewContractAddr :: TAddress (MText, Maybe MText) -- Used in registry dao `genLookupRegistryEntrypoint`
+  }
+
+-- | We cannot get the address of guardian and governance contract
+-- until those contracts originating in the nettest.
+-- As a result, we generate a function that accepts those addresses instead.
+type MkPropertyTestInput = PropertyTestArg -> PropertyTestInput
+
+-- Needed by `forall`
+instance Show (PropertyTestArg -> PropertyTestInput) where
+  show _ = "<MkPropertyTestInput>"
+
+-- | A type for `genPropose` which is used by registry/treasury dao
+type MkGenPropose =
+     Address
+  -> Address
+  -> Address
+  -> GeneratorT (Address -> Address -> (Parameter, Natural, ProposalKey))
+
+type MkGenCustomCalls = GeneratorT (PropertyTestArg -> [CallCustomParam])
+
+-- | Values required to generate `MkPropertyTestInput`
+data MkPropertyTestOption = MkPropertyTestOption
+  { mptoMkPropose :: MkGenPropose
+  , mptoMkCustomCalls :: MkGenCustomCalls
+  , mptoModifyFs :: (FullStorage -> FullStorage)
```

Comment:
Basically, this is a function that modify the generated `FullStorage`. It is mainly used by Treasury/Registry dao SMT which added the `proposalCheck` `decisionLambda` `sExtra`, `rejectedProposalSlashValue` and `Custom entrypoints` from `registryDAO/treasuryDAO_storage.tz`

Even though, our haskell model does not need this since it uses its own haskell implementation, the ligo version still need it.

Reply:
@rinn7e I see, but in unit tests we load the initial storage of the contract from the produced files, couldn't we do the same here? 

AFAIU there is only a single use of this function, at the beginning of test running.

Reply:
Yeah, we do the same as the unit tests: https://github.com/tqtezos/baseDAO/pull/282/files#diff-e3945df4494b4f617adb25abb6bfdc4c15042298f870519743989c64fde2dcc1R39

But in this case, we wouldn't want to use the whole storage, since we have `genFullStorage`. That's why we need a way to modify this generated storage to have proper lambdas. 

We can move this "reading storage" inside the generator directly, so we don't need this `modifyFs`. Let me know what you think.